swap/peer.go

```diff
@@ -0,0 +1,223 @@
+// Copyright 2019 The go-ethereum Authors
+// This file is part of the go-ethereum library.
+//
+// The go-ethereum library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Lesser General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// The go-ethereum library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU Lesser General Public License for more details.
+//
+// You should have received a copy of the GNU Lesser General Public License
+// along with the go-ethereum library. If not, see <http://www.gnu.org/licenses/>.
+
+package swap
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"math/big"
+
+	"github.com/ethereum/go-ethereum/common"
+
+	"github.com/ethereum/go-ethereum/accounts/abi/bind"
+	cswap "github.com/ethersphere/swarm/contracts/swap"
+	"github.com/ethersphere/swarm/log"
+	"github.com/ethersphere/swarm/p2p/protocols"
+)
+
+// ErrDontOwe indictates that no balance is actially owned
+var ErrDontOwe = errors.New("no negative balance")
+
+// Peer is a devp2p peer for the Swap protocol
+type Peer struct {
+	*protocols.Peer
+	swap               *Swap
+	backend            cswap.Backend
+	beneficiary        common.Address
+	contractAddress    common.Address
+	lastReceivedCheque *Cheque
+}
+
+// NewPeer creates a new swap Peer instance
+func NewPeer(p *protocols.Peer, s *Swap, backend cswap.Backend, beneficiary common.Address, contractAddress common.Address) *Peer {
+	return &Peer{
+		Peer:            p,
+		swap:            s,
+		backend:         backend,
+		beneficiary:     beneficiary,
+		contractAddress: contractAddress,
+	}
+}
+
+// handleMsg is for handling messages when receiving messages
+func (sp *Peer) handleMsg(ctx context.Context, msg interface{}) error {
+	switch msg := msg.(type) {
+
+	case *EmitChequeMsg:
+		return sp.handleEmitChequeMsg(ctx, msg)
```

Comment:
Related: https://github.com/ethersphere/swarm/pull/1505#discussion_r298107881 
Maybe a team-wide (`core`) issue should be opened about this.

Although I basically agree, I wonder if in this case it is a good idea. It might be appropriate to block in only this case in order to not mess up accounting. Mess up in this context though would mean that the balance could be manipulated by some other message before the cheque has been completely handled. I see only limited damage by this though as it just means that the cheque amount would be discounted from whatever happens to be the balance at the point of processing if another message was faster in changing the balance. The balance itself is locked of course. 

In short, I think I can agree to make it async but would like to have confirmation though by @zelig @mortelli @vojtechsimetka on this.

Reply:
you probably mean "...habit of treating handlers **async**?

Reply:
I think it would be good to allow `handleEmitChqueMsg` to be executed concurrently, especially since this is a handler function which reacts on incoming messages (imagine what happens if this peer is being flooded by incoming sendCheque messages). 

I only see a possibility of messing up with accounting when the node is very close to the disconnect threshold and while executing the handleEmitChequeMsg goes above this threshold and disconnects... This is, however, an edge case and will not happen when the counterparty node sends his cheques in time.